[FIX] website_slides: fix slides/channel progression while merging partners

pyduf opened the pull request:
The automatic partner merge doesn't work when the source and the destination
partners have done common slides/channels. This fixes the problem.

How to reproduce the problem:
- enroll 2 different partners to the same course
- complete partially the course with the 2 partners with at least one common
slide
- merge the 2 partners
- only what has been done by the destination partner is kept

The same applies with channels, if your partners to merge have done common
channel, only channel done by the destination partner will be kept.

Technical note: there is a uniq constraint on slide.slide.partner on (slide_id,
partner_id) preventing the automatic partner merge to actually merge what has
been done by the partners. Indeed, while assigning slide done by the source
partners to the destination partner, it create a constraint violation exception
and in that case the automatic procedure only keep the slide done by the
destination partner discarding all that has been done by the source partners.

[IMP] base: allow to perform tasks before and/or after partner merge

To allow modules to perform tasks before and/or after a partner merge, we
expose a method _merge_internal that is called when merging partner and that
can be overridden.

[IMP] website_slides: add partner merge test specific to website_slide

The new tests check that merging partners with overlapping course completion
(common slides completed/channel enrolled) assign the union of what has been
done by the source and destination partners to the destination partner.

It also tests that the karma is adequately adjusted.

Task-3167160


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/114840).

flch-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,93 @@
+        current_id_kept = None
+        for record in Model.search_read([('partner_id', 'in', (src_partners + dst_partner).ids)],
+                                        [uniq_field, 'partner_id'],
+                                        order=uniq_field + ((',' + preferences) if preferences else '')):
```
Why not set `preferences` to `''` by default instead of requiring to call with `''` as it is a real use case we have below?

(also, fewer parentheses with
```suggestion
                                        order=uniq_field + (f', {preferences}' if preferences else ''):
```
)

flch-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,93 @@
+
+    @api.model
+    def _update_foreign_keys(self, src_partners, dst_partner):
+        # Due to cascade deletion, we do the full merge here and don't delegate it partially to _update_foreign_keys
```
```suggestion
        """Overridden to avoid cascade deletions during the base method's merging procedure."""
```

flch-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,93 @@
+        self._merge_partner_with_preferences(
+            src_partners, dst_partner,
+            'slide.slide.partner', 'slide_id', 'completed DESC, vote DESC, quiz_attempts_count ASC')
+        # We keep record with dst_partner to avoid cascade deletion of slide.slide.partner
```
```suggestion
```

flch-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,93 @@
+    @api.model
+    def _merge_partner_with_preferences(self, src_partners, dst_partner,
+                                        res_model, uniq_field, preferences, keep_dst_partner=False):
+        """ Merge records of "table" for the given partners following the given preferences.
```
"table" is `res_model`?

flch-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,93 @@
+
+        :param recordset src_partners: source partners
+        :param recordset dst_partner: destination partner
+        :param str res_model: model name
+        :param str uniq_field: column name of the unique key per partner
```
If I understood correctly, this methods handle m2m relationships. I would specify it in the docstring and maybe rename a few things like 
```suggestion 
        :param str res_model: many to many relationship model
        :param str other_record_field: field name of the unique "other" record per partner
```

flch-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,93 @@
+                                        order=uniq_field + ((',' + preferences) if preferences else '')):
+            uniq_id = record[uniq_field][0]
+            if uniq_id == prev_uniq_id:  # duplicate to remove
+                if keep_dst_partner and record['partner_id'][0] == dst_partner.id:
```
Isn't it easier to do a separate flow that simply searches and unlinks records not linked to dst_partner if this flag is set? (And cleaner when we see the rest of the algorithm).

flch-odoo commented on addons/website_slides/tests/test_partner_merge.py:
```diff
@@ -0,0 +1,146 @@
+    def test_merge_2_partners(self):
+        # for partner_ids_to_merge in permutations([u.partner_id.id for u in self.user_students[:2]]):
+        for partner_ids_to_merge in permutations([u.partner_id.id for u in self.user_students[:2]]):
+            with self.subTest(partner_to_merge=partner_ids_to_merge):
```
```suggestion
            with self.subTest(partner_ids_to_merge=partner_ids_to_merge):
```

flch-odoo reviewed (commented):
Nice work! 
It was a bit hard to read at first however. So, after a focused read, a few things came in mind to try and simplify it a little :)

pyduf commented:
> Nice work! It was a bit hard to read at first however. So, after a focused read, a few things came in mind to try and simplify it a little :)

Hello @flch-odoo, thanks for your suggestions. I have updated the code accordingly.

flch-odoo reviewed (commented):
A few details, but validation-wise, this is OK to me :)

std-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,94 @@
+        Model_sudo = self.env[res_model].sudo()
+        records_to_delete = list()
+        grouped_records = groupby(Model_sudo.search_read([('partner_id', 'in', (src_partners + dst_partner).ids)],
+                                                         [other_record_field, 'partner_id'],
```
I have to scroll horizontally to read it :/ can you do something like this ?
```py
        grouped_records = groupby(Model_sudo.search_read(
            [('partner_id', 'in', (src_partners + dst_partner).ids)],
            [other_record_field, 'partner_id'],
            order=other_record_field + (f', {order}' if order else '')),
            key=lambda r: r[other_record_field][0],
        )
```

std-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,94 @@
+                    records_to_delete += others[1:]
+        else:
+            for _, records in grouped_records:
+                records_to_delete += islice(records, 1, None)
```
if you use our own groupby
```
from odoo.tools.misc import groupby
```

you can just do
```
records_to_delete += records[1:]
```

(and you remove one import)

std-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,94 @@
+                                  key=lambda r: r[other_record_field][0])
+        if prefer_dst_partner:
+            for _, records in grouped_records:
+                dst_partner_records, others = partition(lambda r: r['partner_id'][0] == dst_partner.id, records)
```
same here, I need to scroll :/

std-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,94 @@
+        with dst_partner if present instead of following the preferences
+        """
+        Model_sudo = self.env[res_model].sudo()
+        records_to_delete = list()
```
```suggestion
        records_to_delete = []
```

(warning from: https://pypi.org/project/flake8-comprehensions/ )

std-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,94 @@
+        :param recordset dst_partner: destination partner
+        :param str res_model: many-to-many relationship model linking res_partner and another
+            model through the other_record_field
+        :param str other_record_field: field name of the unique "other" record per partner
```
the name `other_record_field` can be more explicit I think, maybe `uniq_field` or something else ? (would like to know, just by reading the name, that we will keep only 1 record / value of this field if you see what I mean)

std-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,94 @@
+                dst_partner_records, others = partition(lambda r: r['partner_id'][0] == dst_partner.id, records)
+                if dst_partner_records:
+                    records_to_delete += others
+                    records_to_delete += dst_partner_records[1:]
```
about this line, are you sure ?
```suggestion
```

we already remove `others`, do we need to remove `dst_partner_records[1:]` as well ?
(the test is not crashing if I comment this line)


same question about
```py
records_to_delete += records[1:]
```

if you group  partner with a other partner that has zero slide.partner, it will still remove some

std-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,94 @@
+
+        # Recompute the completion
+        for slide_partner_relation in SlidePartner_sudo.search([('partner_id', '=', dst_partner.id)]):
+            slide_partner_relation._recompute_completion()
```
`_recompute_completion` can be called in batch, is there a reason to iterate ?

if it's because of the domain used in `_recompute_completion` maybe we can improve it with OR conditions ? (or adding an ensure_one to be sure it's called correctly)

std-odoo commented on addons/website_slides/wizard/base_partner_merge.py:
```diff
@@ -0,0 +1,94 @@
+        # Karma correction: Karma has been impacted incorrectly by the recompute of the completion
+        earned_karma = all_channels._get_earned_karma([dst_partner.id])[dst_partner.id]
+        for user, partner_id in user_partner_id_lst:
+            delta_karma = earned_karma - init_earned_karma_by_partner_id.get(partner_id, 0)
+            user.karma = user_karma[user] + delta_karma
```
so the karma will stay the same as before ? (even if the other partner had karma)

shouldn't we add the karma instead ? (the new karma is the sum of all karma)

std-odoo replied:
ah no ok, we will have 2 users that will share the same partner after the merge Oo
that's super weird no ? (it should be impossible to merge partner if they have a user ?)

std-odoo reviewed (commented):
@pyduf Hi :) 

Good work :) I have a some proposition and questions

Also for this type of thing, it would be nice to have a 100% coverage in the test (see below, but some part of the code are not tested)

Cheers

std-odoo commented on addons/website_slides/tests/test_partner_merge.py:
```diff
@@ -0,0 +1,146 @@
+
+    def test_merge_2_partners(self):
+        # for partner_ids_to_merge in permutations([u.partner_id.id for u in self.user_students[:2]]):
+        for partner_ids_to_merge in permutations([u.partner_id.id for u in self.user_students[:2]]):
```
I think all run after the first one will be useless
![screenshot_spect](https://user-images.githubusercontent.com/54316226/234599368-b2c91ca0-6b88-4e36-8ded-39abd1d6ba50.png)

pyduf commented:
Hello @std-odoo, 

Thank you for your valuable comments. I have implemented the changes for most of them. It remains 2 points for which I will come back later: 
- covering all the cases of _merge_partners_for_unique_m2m in the test
- as you will see in the comment, following a discussion with Aurélien, remove the karma adjustment (but keep this version in case we need it anyway) and try to modify the method _recompute_completion so that it doesn’t modify the karma while merging partner.

I will ping you when it is one.

pyduf commented:
> Hello @std-odoo,
> 
> Thank you for your valuable comments. I have implemented the changes for most of them. It remains 2 points for which I will come back later:
> 
> * covering all the cases of _merge_partners_for_unique_m2m in the test
> * as you will see in the comment, following a discussion with Aurélien, remove the karma adjustment (but keep this version in case we need it anyway) and try to modify the method _recompute_completion so that it doesn’t modify the karma while merging partner.
> 
> I will ping you when it is one.

Hello @std-odoo , 

Actually, with the correction of the test, the method _merge_partners_for_unique_m2m is fully covered. I have also simplified the merging following the decision to not update the karma anymore.
